### PR TITLE
[TC-5834] Added support for deferred app links

### PR DESCRIPTION
**Pull request by kopiro**
https://jira.appcelerator.org/browse/TC-5834

### CODE REVIEW

File: ios/Classes/FacebookModule.m [hunk 1]
```diff
@@ -636,6 +636,35 @@ -(void)requestWithGraphPath:(id)args
...
+
+    TiThreadPerformOnMainThread(^{
+        [FBSDKAppLinkUtility fetchDeferredAppLink:^(NSURL *url, NSError *error) {
+            NSDictionary* returnedObject;
+
+            if (url) {
+                returnedObject = [[NSDictionary alloc] initWithObjectsAndKeys:[url absoluteURL],@"url", nil];
+            } else {
+                if (error) {
+                    NSLog(@"Received error while fetching deferred app link %@", error);
```
Comment (hansemannn) [inline ios/Classes/FacebookModule.m]: Please prefix the log with `[ERROR]` to display it in the console.

Comment (kopiro) [inline ios/Classes/FacebookModule.m]: @hansemannn done :)

**Comment by cheekiatng** [comment]
Thanks for the PR @kopiro ! Our team will take a look at this!

File: README.md [hunk 2]
```diff
@@ -88,7 +88,7 @@ If you choose to enable it, you have to set the following keys and values in tia
 
 On the android platform, in tiapp.xml or AndroidManifest.xml you must declare the following inside the \<application\> node 
 ```xml
-<activity android:name="com.facebook.LoginActivity" android:theme="@android:style/Theme.Translucent.NoTitleBar" android:label="YourAppName"/>
+<activity android:name="com.facebook.FacebookActivity" android:theme="@android:style/Theme.Translucent.NoTitleBar" android:label="YourAppName"/>
```
Comment (ashcoding) [inline README.md]: This seems to be a duplicate of line 92? Perhaps this can be removed instead.

**Comment by ashcoding** [comment]
Please remove the the file `build.properties` at https://github.com/kopiro/ti.facebook/blob/PR-TC-5834/android/build.properties from the PR. :) Thank you.

**Comment by ashcoding** [comment]
@kopiro Would you have an example on how this is used or how you have used this in your own app?
This would help in making the documentation and testing more complete.

**Comment by kopiro** [comment]
@ashcoding in the index.js, when the UI is complete:

``` js
FB.fetchDeferredAppLink(function(e) {
   if (e.url != null) {
      dispatchInternalRoute(e.url);
   }
});
```

`dispatchInternalRoute` is a function that analyze the incoming url.

I use my own framework (https://github.com/CaffeinaLab/Trimethyl) to handle this (_Router_ class).

**Comment by ashcoding** [comment]
@kopiro, could you address the comments and remove the parts that was mentioned? That would be awesome if you could! :)

**Comment by kopiro** [comment]
@ashcoding voila: https://github.com/kopiro/ti.facebook/commit/e575f0d5a31af933f40e558ddb5e892a58b9c252

**Comment by hansemannn** [comment]
Hey @kopiro, I would like to merge your merges this week. Can you submit a new PR that only adds the new methods + docs? The included Facebook SDK is already covered by the latest ti.facebook release and the [adjusted build script](https://github.com/appcelerator-modules/ti.facebook/pull/39/files#diff-a69a274edae3910fa5c8e8ee83b90350R72) could mess-up our build process. Submitting a new & clean PR against master would be awesome.

**Comment by kopiro** [comment]
@hansemannn done!
Let me know if it's 200 OK :P

**Comment by hansemannn** [comment]
CR looks good! Can you add the method inside the `apidoc` directory, as well as indent the example (the callback)?

**Comment by kopiro** [comment]
@hansemannn since SDK version?

**Comment by hansemannn** [comment]
5.4.0

**Comment by kopiro** [comment]
@hansemannn I refactored adding the `success` property.

**Comment by kopiro** [comment]
@hansemannn 

```
cp: build.properties: No space left on device
cat: build.properties: No such file or directory
```

In the build.

**Comment by hansemannn** [comment]
I just refactored the iOS-implementation and fixed some issues related to Dictionaries that might (also) caused problems. Give it a try:

``` objc
-(void)fetchDeferredAppLink:(id)args
{
    ENSURE_TYPE(args, NSArray);
    ENSURE_TYPE([args objectAtIndex:0], KrollCallback);
    KrollCallback *callback = [args objectAtIndex:0];

    TiThreadPerformOnMainThread(^{
        [FBSDKAppLinkUtility fetchDeferredAppLink:^(NSURL *url, NSError *error) {
            NSMutableDictionary* event = [NSMutableDictionary dictionaryWithDictionary:@{@"success": NUMBOOL(url != nil)}];
            if (url) {
                [event setValue:url forKey:@"url"];
            } else {
                NSString *errorString = @"An error occurred. Please try again.";
                if (error != nil) {
                    errorString = [[error userInfo] objectForKey:FBSDKErrorLocalizedDescriptionKey];
                }
                [event setValue:errorString forKey:@"error"];
            }

            KrollEvent * invocationEvent = [[KrollEvent alloc] initWithCallback:callback eventObject:event thisObject:self];
            [[callback context] enqueue:invocationEvent];
            [invocationEvent release];
            [event release];
        }];
    }, YES);
}
```

File: android/src/facebook/TiFacebookModule.java [hunk 3]
```diff
@@ -611,6 +612,29 @@ public void onActivityResult(Activity activity, int requestCode, int resultCode,
...
+			@Override
+			public void onDeferredAppLinkDataFetched(AppLinkData appLinkData) {
+				KrollDict data = new KrollDict();
+
+				if (appLinkData == null) {
+					data.put("success", false);
+					data.put("error", "An error occurred. Please try again.");
+				} else {
+					data.put("success", true);
+					data.put("error", null);
```
Comment (hansemannn) [inline android/src/facebook/TiFacebookModule.java]: If no error occurred, we should not add the "error" key

File: README.md [hunk 4]
```diff
@@ -380,6 +380,22 @@ To create a Messenger button, call the `createMessengerButton` method. Example:
...
+------------------
+
+Deferred deep linking allows you to send people to a custom view after they installed your app via the app store.
+
+You can simply call fetchDeferredAppLink on startup to open eventually incoming app links.
+
+```javascript
+var fb = require('facebook');
+fb.fetchDeferredAppLink(function(e) {
+    if (e.url) {
```
Comment (hansemannn) [inline README.md]: Because we are checking the url-existence internally, we should just check `e.success` here like done in other API's.

**Comment by kopiro** [comment]
@hansemannn Wait, it crash. :(
I'm inspecting why.

**Comment by hansemannn** [comment]
@kopiro iOS? Let me see what's happening there. Also please pull the latest master to be in sync, thanks!

**Comment by kopiro** [comment]
Pulled from master and merged! @hansemannn

File: android/example/facebook_messenger.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+exports.window = function(value){
```
Comment (hansemannn) [inline android/example/facebook_messenger.js]: Why is this file added? The messenger API's are not supported on Android, yet. Probably merge conflicts?

**Comment by hansemannn** [comment]
Also please remove the binary from the pull, we courage zipped packaged in the Releases tab 😊

**Comment by kopiro** [comment]
@hansemannn done!

**Comment by hansemannn** [comment]
Don't forget to push that change

**Comment by kopiro** [comment]
I pushed :)
(forced :D)

**Comment by hansemannn** [comment]
Approved!
